### PR TITLE
docs: Convert ARCOptimization.rst to Markdown

### PR DESCRIPTION
**Pull request by MaxDesiatov**
Currently docs are a mix of reStructuredText and Markdown and unifying them as Markdown would make more sense long term, as far as I understand. This converts `ARCOptmization.rst`, initially converted with [Pandoc](https://pandoc.org/) with manual cleanup added after the fact.

### CODE REVIEW

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

File: docs/ARCOptimization.md [hunk 1]
```diff
@@ -0,0 +1,947 @@
...
+-   `strong_release`
+-   `strong_retain_unowned`
+-   `unowned_retain`
+-   `unowned_release`
+-   `load_weak`
+-   `store_weak`
+-   `fix_lifetime`
+-   `mark_dependence`
+-   `is_unique`
+-   `copy_block`
```
Comment (varungandhi-apple) [inline docs/ARCOptimization.md:29]: You can remove the additional spacing between `-` and the backtick.

File: docs/ARCOptimization.md [hunk 2]
```diff
@@ -0,0 +1,947 @@
...
+argument type:
+
+> -   Native object types are directly checked by reading the strong
+>     reference count: (Builtin.NativeObject, known native class
+>     reference)
+> -   Objective-C object types require an additional check that the
+>     dynamic object type uses native Swift reference counting: (unknown
+>     class reference, class existential)
+> -   Bridged object types allow the dynamic object type check to be
+>     bypassed based on the pointer encoding: (Builtin.BridgeObject)
```
Comment (varungandhi-apple) [inline docs/ARCOptimization.md:318]: Maybe this was accidentally marked as a quote? Seems a bit odd... It seems like it is stating the rules, and not quoting rules stated elsewhere.

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

**Comment by benrimmington** [comment]
There's also a link in [docs/README.md](https://github.com/apple/swift/blame/8a0e7062cc3fff11bdfcda37097eabbbe5e599cb/docs/README.md#L307).

**Comment by MaxDesiatov** [comment]
Thanks, that's fixed now.

File: docs/ARCOptimization.md [hunk 3]
```diff
@@ -0,0 +1,941 @@
+# ARC Optimization for Swift
+
+*TODO*: This is an evolving document on ARC optimization in the Swift compiler.
+Please extend it.
+
+## Terms
```
Comment (benrimmington) [inline docs/ARCOptimization.md:6]:
```suggestion
## Contents

* [Terms](#terms)
* [Reference Counting Instructions](#reference-counting-instructions)
* [Memory Behavior of ARC Operations](#memory-behavior-of-arc-operations)
* [ARC and Copying](#arc-and-copying)
* [RC Identity](#rc-identity)
  * [Definitions](#definitions)
  * [Contrasts with Alias Analysis](#contrasts-with-alias-analysis)
  * [What is `retain_value` and why is it important?](#what-is-retain_value-and-why-is-it-important)
  * [Conversions](#conversions)
  * [ARC and Enums](#arc-and-enums)
* [Copy-On-Write Considerations](#copy-on-write-considerations)
  * [`is_unique` instruction](#is_unique-instruction)
  * [`Builtin.isUnique`](#builtinisunique)
* [Semantic Tags](#semantic-tags)
  * [`programtermination_point`](#programtermination_point)
* [Unreachable Code and Lifetimes](#unreachable-code-and-lifetimes)
* [ARC Sequence Optimization](#arc-sequence-optimization)
* [ARC Loop Hoisting](#arc-loop-hoisting)
  * [Abstract](#abstract)
  * [Loop Canonicalization](#loop-canonicalization)
  * [Motivation](#motivation)
  * [Correctness](#correctness)
  * [Compensating Early Exits for Lost Dynamic Reference Counts](#compensating-early-exits-for-lost-dynamic-reference-counts)
  * [Uniqueness Check Complications](#uniqueness-check-complications)
* [Deinit Model](#deinit-model)

## Terms
```

**Review by benrimmington — COMMENTED** [review]
I did a quick side-by-side comparison.
The only thing I noticed was a missing table of contents.

File: docs/ARCOptimization.md (hunk 3 above)
Comment (MaxDesiatov) [reply docs/ARCOptimization.md:6]: Thanks! For some reason skipped this section...

**Review by benrimmington — APPROVED** [review]

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test and merge

**Review by varungandhi-apple — APPROVED** [review]

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

**Comment by benrimmington** [comment]
@swift-ci Please smoke test Linux platform

**Comment by MaxDesiatov** [comment]
@swift-ci please test Windows platform

**Comment by Teemperor** [comment]
```
writing output... [ 98%] proposals/valref
writing output... [100%] weak


Warning, treated as error:
/Users/teemperor/swift/master2/swift/docs/SIL.rst:4247:undefined label: arcopts.is_unique
[264/1631][ 16%][26.504s] Building CXX object stdlib/public/runtime/CMakeFiles/swiftRuntime-macosx-arm64e.dir/Metadata.cpp.o
ninja: build stopped: subcommand failed.
ERROR: command terminated with a non-zero exit status 1, aborting
```

This breaks the doc compilation which in turns breaks the complete build. Please revert, thank you.

**Comment by benrimmington** [comment]
> This breaks the doc compilation which in turns breaks the complete build. Please revert, thank you.

@Teemperor Possible fix: apple/swift#36131

**Pull request merged**